## ABRIMOS EL DATASET

In [1]:
import pandas as pd

In [2]:
ls

 NIRSdata.csv                 ProyectoFinal-Objetivos.pdf
'Proyecto Final - IA.ipynb'   README.md


In [3]:

data = pd.read_csv('NIRSdata.csv', encoding='latin1')

/tmp/ipykernel_85096/1622662162.py:1: DtypeWarning: Columns (3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,21,22,23,25,26,27,28,29,30,31,32,35,36,37,38,39,42,47,49) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv('NIRSdata.csv', encoding='latin1')


In [4]:
data

Unnamed: 0 CODIGO_MUESTRA  2X        Codigos Codigo_de_muestra  \
0              1  LQAS19-002703 NaN  LQAS19-002703     LQAS19-002703   
1              2  LQAS19-002703 NaN  LQAS19-002703     LQAS19-002703   
2              3  LQAS19-002704 NaN  LQAS19-002704     LQAS19-002704   
3              4  LQAS19-002704 NaN  LQAS19-002704     LQAS19-002704   
4              5  LQAS19-002705 NaN  LQAS19-002705     LQAS19-002705   
...          ...            ...  ..            ...               ...   
6173        6174  LQAS19-008827 NaN            NaN               NaN   
6174        6175  LQAS19-008828 NaN            NaN               NaN   
6175        6176  LQAS19-008829 NaN            NaN               NaN   
6176        6177  LQAS19-008830 NaN            NaN               NaN   
6177        6178  LQAS19-008831 NaN            NaN               NaN   

         Fecha_de_registro    Cultivo Sistema_cultivo Profundidad  \
0     4/06/2019 3:22 p. m.   Maracuya       No Indica   0 a 30 cm   
1     4/06/2019 3:22 p. m.   Maracuya       No Indica   0 a 30 cm   
2     4/06/2019 3:22 p. m.   Maracuya       No Indica   0 a 30 cm   
3     4/06/2019 3:22 p. m.   Maracuya       No Indica   0 a 30 cm   
4     4/06/2019 3:22 p. m.  Guanabana     Monocultivo   0 a 30 cm   
...                    ...        ...             ...         ...   
6173                   NaN        NaN             NaN         NaN   
6174                   NaN        NaN             NaN         NaN   
6175                   NaN        NaN             NaN         NaN   
6176                   NaN        NaN             NaN         NaN   
6177                   NaN        NaN             NaN         NaN   

     Estado_cultivo  ...    2414_5      2423    2431_5      2440    2448_5  \
0       Establecido  ...  0.470079  0.474854  0.480904  0.487413  0.494017   
1       Establecido  ...  0.480973  0.485957  0.492129  0.498875  0.505900   
2       Establecido  ...  0.523511  0.528787  0.535298  0.542274  0.549406   
3       Establecido  ...  0.570260  0.575961  0.582749  0.590119  0.597509   
4       Establecido  ...  0.531322  0.535443  0.540683  0.546291  0.551770   
...             ...  ...       ...       ...       ...       ...       ...   
6173            NaN  ...       NaN       NaN       NaN       NaN       NaN   
6174            NaN  ...       NaN       NaN       NaN       NaN       NaN   
6175            NaN  ...       NaN       NaN       NaN       NaN       NaN   
6176            NaN  ...       NaN       NaN       NaN       NaN       NaN   
6177            NaN  ...       NaN       NaN       NaN       NaN       NaN   

          2457    2465_5      2474    2482_5      2491  
0     0.500904  0.508063  0.514994  0.521809  0.527157  
1     0.513012  0.520427  0.527738  0.534733  0.540491  
2     0.556501  0.564125  0.571468  0.578451  0.584249  
3     0.605052  0.612833  0.620518  0.627860  0.633891  
4     0.557568  0.563411  0.569355  0.575130  0.579539  
...        ...       ...       ...       ...       ...  
6173       NaN       NaN       NaN       NaN       NaN  
6174       NaN       NaN       NaN       NaN       NaN  
6175       NaN       NaN       NaN       NaN       NaN  
6176       NaN       NaN       NaN       NaN       NaN  
6177       NaN       NaN       NaN       NaN       NaN  

[6178 rows x 297 columns]

## ANALISIS EXPLORATORIO DE LOS DATOS: ¿CÓMO Y QUÉ VIENE EN LOS DATOS?

El dataset esta codificado en español utilizando tildes y demás, esto causa problemas de lectura en los tipos de datos. 

Voy a separar las columnas de datos de absorbancia de las caracteristicas del suelo.

In [5]:
import numpy as np

In [6]:
soil_properties_columns = data.iloc[:, 0:50]
absorbance_columns = data.iloc[:, 50:]

In [7]:
absorbance_columns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6178 entries, 0 to 6177
Columns: 247 entries, 400 to 2491
dtypes: float64(247)
memory usage: 11.6 MB


Podemos ver que el dataset de absorbancia desde 400nm hasta 2491nm son 247 columnas en total y 6178 filas. Todas ellas del tipo 'float64'. Podemos decir que este dataset esta decente por el momento, parece que el dateset esta completo y correcto, sin emabrgo eso lo verificaremos más adelante. 

Ahora para el dataset de caracteristicas del suelo:

In [8]:
soil_properties_columns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6178 entries, 0 to 6177
Data columns (total 50 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Unnamed: 0                        6178 non-null   int64  
 1   CODIGO_MUESTRA                    6178 non-null   object 
 2   2X                                0 non-null      float64
 3   Codigos                           3673 non-null   object 
 4   Codigo_de_muestra                 3671 non-null   object 
 5   Fecha_de_registro                 3671 non-null   object 
 6   Cultivo                           3671 non-null   object 
 7   Sistema_cultivo                   3671 non-null   object 
 8   Profundidad                       3671 non-null   object 
 9   Estado_cultivo                    3671 non-null   object 
 10  Unidad_edad_cultivo               3671 non-null   object 
 11  Topografia                        3668 non-null   object 
 12  Drenaj

Acá podemos ver bien cada columna y que hace falta por arreglar  o limpiar.

- Quitaré las columnas que tienen '0 non-null' o que todas sus filas son NaN, Tabien quitare las columnas que tengan '<200 non-null'

Esto lo hago porqué al tener 6178 entradas , tener solo 200 filas YO lo considero que no es representativo de la población al ser ~3% del total de los datos únicamente para hacer inferencia del 97% restante.

- Eliminare las columnas que no presentan información , que son metadatos que no sirven para hacer inferencias

In [9]:
soil_properties_columns = soil_properties_columns.drop(columns=['2X','2y','OBSERVACIONES','CICA','humedad','textura','Arcilla','Arena','Limo'])
soil_properties_columns = soil_properties_columns.drop(columns=['id'])


El principal problema es que muchas columnas (MAgnesio, Hierro, Fosforo...) NO estan en tipo float64 sino en tipo object.
- Dividire el dataset de 'soil_properties_columns' entre propiedades y metadatos
- Iniciaré mirando algunas filas del dataset para determinar en cada columna el porqué de los valores Nulos si se debe a 'NaN' o se debe a casos de formato erroneo que contiene información.

In [10]:
selected_columns=['Unnamed: 0','CODIGO_MUESTRA','Codigos','Codigo_de_muestra','Fecha_de_registro','sampid_x','FECHA_DEL_SCAN','sampid_x','sampid_y']
metadata_soil = soil_properties_columns[selected_columns]
remaining_columns = [col for col in soil_properties_columns.columns if col not in selected_columns]
properties_soil = soil_properties_columns[remaining_columns]



In [11]:
metadata_soil.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6178 entries, 0 to 6177
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Unnamed: 0         6178 non-null   int64 
 1   CODIGO_MUESTRA     6178 non-null   object
 2   Codigos            3673 non-null   object
 3   Codigo_de_muestra  3671 non-null   object
 4   Fecha_de_registro  3671 non-null   object
 5   sampid_x           6178 non-null   object
 6   FECHA_DEL_SCAN     3915 non-null   object
 7   sampid_x           6178 non-null   object
 8   sampid_y           6178 non-null   object
dtypes: int64(1), object(8)
memory usage: 434.5+ KB


In [12]:
properties_soil.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6178 entries, 0 to 6177
Data columns (total 32 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Cultivo                           3671 non-null   object 
 1   Sistema_cultivo                   3671 non-null   object 
 2   Profundidad                       3671 non-null   object 
 3   Estado_cultivo                    3671 non-null   object 
 4   Unidad_edad_cultivo               3671 non-null   object 
 5   Topografia                        3668 non-null   object 
 6   Drenaje                           3671 non-null   object 
 7   Riego                             3671 non-null   object 
 8   Prod_cultivo                      3671 non-null   object 
 9   Req_recomend_fertilizacion        3671 non-null   object 
 10  Departamento                      3671 non-null   object 
 11  Finca                             3640 non-null   object 
 12  Munici

In [13]:
unique_classes = properties_soil['Estado_cultivo'].unique()
print('Columna Estado_cultivo: ', unique_classes)

for column in properties_soil.columns:
    unique_classes = properties_soil[column].unique()
    print(f"Columna '{column}':")
    print(unique_classes)
    print()

Columna Estado_cultivo:  ['Establecido' 'Por establecer' nan 'No indica' '0']
Columna 'Cultivo':
['Maracuya' 'Guanabana' 'Cacao' 'Cania' 'Chontaduro' nan 'Maderable'
 'No Indica' 'Arroz' 'Cania Azucarera' 'Uva' 'Aji' 'Papaya' 'Aguacate'
 'Granadilla' 'Cafe' 'Ma�z' 'Lim�n' 'Tomate' 'Mango'
 'Hortalizas Varias' 'Pi�a' 'Naranja' 'Mandarina' 'Rambutan' '0'
 'Frutales Varios' 'Guayaba' 'Pasto' 'C�tricos' 'Palma de Aceite'
 'Berenjena' 'Kikuyo' 'Pasto Brachiaria' 'Sacha Inchi' 'Cania Panelera'
 'Papa' 'Pl�tano' 'Ajo' 'Lulo' 'Ma�z Forrajero' 'Mel�n' 'S�bila'
 'Uchuva' 'Mara��n' 'Pimienta' 'Pitahaya' 'Habichuela' 'Cannabis'
 'Fresa' 'Yuca' 'Alfalfa' 'Mora' 'Lechuga' 'Arveja' 'Rosa' 'Romero'
 'Zanahoria' 'Cilantro' 'Follajes' 'Patilla' 'Lima' 'Frijol' 'Ahuyama'
 'Mangostino' 'Caucho' 'Matarraton' 'Ar�ndano' 'Guadua' 'Gulupa'
 'Pasto Par�' 'Soya' 'Banano' 'Tangelo' 'Menta' 'Higuerilla'
 'Pasto Angleton' 'Pasto Estrella' 'Cebolla de Rama' 'Batata' 'Trigo'
 'Fique' 'Silvopa

In [14]:
#NOS ENCARGAMOS DE LAS COLUMNAS DE TIPO 'float64'
import re

specific_columns=['pH','Materia_organica','CICE', 'Fosforo','Azufre','Acidez', 'Position','Boro','Aluminio', 'Calcio', 'Magnesio', 'Potasio', 'Sodio', 'Conductividad', 'Hierro', 'Cobre','Manganeso','Zinc']

for column in specific_columns:
    if properties_soil[column].dtype == 'object':
        properties_soil[column] = properties_soil[column].str.replace(r'^[<>]([\d.]+)', r'\1', regex=True)
        properties_soil[column] = properties_soil[column].replace('ND', np.nan)
        properties_soil[column]=properties_soil[column].astype(float)
        
properties_soil[specific_columns] = properties_soil[specific_columns].interpolate(method='linear', limit_direction='both')


/tmp/ipykernel_85096/3810821607.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  properties_soil[column] = properties_soil[column].str.replace(r'^[<>]([\d.]+)', r'\1', regex=True)
/tmp/ipykernel_85096/3810821607.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  properties_soil[column] = properties_soil[column].replace('ND', np.nan)
/tmp/ipykernel_85096/3810821607.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = 

In [15]:
#NOS ENCARGAMOS DE LAS COLUMNAS DE TIPO 'object'

replacement_mapping = {np.nan: 'No indica', '0': 'No indica'}
properties_soil['Estado_cultivo'] = properties_soil['Estado_cultivo'].replace(replacement_mapping)

replacement_mapping = {np.nan: 'No indica', '0': 'No indica', 'NO INDICA': 'No indica', 'No Indica': 'No indica'}
properties_soil['Sistema_cultivo'] = properties_soil['Sistema_cultivo'].replace(replacement_mapping)

replacement_mapping = {np.nan: 'No indica', '0': 'No indica', '0_A_32': '0 a 32 cm', '0_A_33': '0 a 33 cm', '0_A_31': '0 a 31 cm'}
properties_soil['Profundidad'] = properties_soil['Profundidad'].replace(replacement_mapping)

replacement_mapping = {'Mese(s)': 'Meses', np.nan: 'No indica', '0': 'No indica'}
properties_soil['Unidad_edad_cultivo'] = properties_soil['Unidad_edad_cultivo'].replace(replacement_mapping)

replacement_mapping = {np.nan: 'No indica', '0': 'No indica', 'NO INDICA': 'No indica', 'No Indica': 'No indica'}
properties_soil['Topografia'] = properties_soil['Topografia'].replace(replacement_mapping)
properties_soil['Drenaje'] = properties_soil['Drenaje'].replace(replacement_mapping)
properties_soil['Riego'] = properties_soil['Riego'].replace(replacement_mapping)
properties_soil['Prod_cultivo'] = properties_soil['Prod_cultivo'].replace(replacement_mapping)
properties_soil['Req_recomend_fertilizacion'] = properties_soil['Req_recomend_fertilizacion'].replace(replacement_mapping)
properties_soil['Departamento'] = properties_soil['Departamento'].replace(replacement_mapping)
properties_soil['Finca'] = properties_soil['Finca'].replace(replacement_mapping)
properties_soil['Municipio'] = properties_soil['Municipio'].replace(replacement_mapping)
properties_soil['Sample_Identification_String_One'] = properties_soil['Sample_Identification_String_One'].replace(replacement_mapping)
properties_soil['Cultivo'] = properties_soil['Cultivo'].replace(replacement_mapping)



/tmp/ipykernel_85096/3141349688.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  properties_soil['Estado_cultivo'] = properties_soil['Estado_cultivo'].replace(replacement_mapping)
/tmp/ipykernel_85096/3141349688.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  properties_soil['Sistema_cultivo'] = properties_soil['Sistema_cultivo'].replace(replacement_mapping)
/tmp/ipykernel_85096/3141349688.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .lo

LISTO¡ Como se observa a continuación hemos limpiado todo el dataset, hemos imputado los valores faltantes utilizando interpolación linear, y hemos completado los valores faltantes de los 'object' mapeando todo valor nan, '0' u erroneo por la clase por defecto 'No indica'. 

In [16]:
properties_soil.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6178 entries, 0 to 6177
Data columns (total 32 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Cultivo                           6178 non-null   object 
 1   Sistema_cultivo                   6178 non-null   object 
 2   Profundidad                       6178 non-null   object 
 3   Estado_cultivo                    6178 non-null   object 
 4   Unidad_edad_cultivo               6178 non-null   object 
 5   Topografia                        6178 non-null   object 
 6   Drenaje                           6178 non-null   object 
 7   Riego                             6178 non-null   object 
 8   Prod_cultivo                      6178 non-null   object 
 9   Req_recomend_fertilizacion        6178 non-null   object 
 10  Departamento                      6178 non-null   object 
 11  Finca                             6178 non-null   object 
 12  Munici

In [17]:
absorbance_columns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6178 entries, 0 to 6177
Columns: 247 entries, 400 to 2491
dtypes: float64(247)
memory usage: 11.6 MB
